internal/jobcontainers/process.go

```diff
@@ -0,0 +1,174 @@
+package jobcontainers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"golang.org/x/sys/windows"
+)
+
+// JobProcess represents a process run in a job object.
+type JobProcess struct {
+	cmd            *exec.Cmd
+	procLock       sync.Mutex
+	stdioLock      sync.Mutex
+	stdin          io.WriteCloser
+	stdout         io.ReadCloser
+	stderr         io.ReadCloser
+	waitBlock      chan struct{}
+	closedWaitOnce sync.Once
+	waitError      error
+}
+
+var _ cow.Process = &JobProcess{}
+
+func newProcess(cmd *exec.Cmd) *JobProcess {
+	return &JobProcess{
+		cmd:       cmd,
+		waitBlock: make(chan struct{}),
+	}
+}
+
+func (p *JobProcess) ResizeConsole(ctx context.Context, width, height uint16) error {
+	return nil
+}
+
+// Stdio returns the stdio pipes of the process
+func (p *JobProcess) Stdio() (io.Writer, io.Reader, io.Reader) {
+	return p.stdin, p.stdout, p.stderr
+}
+
+// Signal sends a signal to the process.
+func (p *JobProcess) Signal(ctx context.Context, options interface{}) (bool, error) {
+	p.procLock.Lock()
+	defer p.procLock.Unlock()
+
+	if p.exited() {
+		return false, errors.New("signal not sent. process already exited")
+	}
+
+	if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, uint32(p.Pid())); err != nil {
+		return false, fmt.Errorf("failed to send signal: %s", err)
+	}
+	return true, nil
+}
+
+// CloseStdin closes the stdin pipe of the process.
+func (p *JobProcess) CloseStdin(ctx context.Context) error {
+	p.stdioLock.Lock()
+	defer p.stdioLock.Unlock()
+	return p.stdin.Close()
+}
+
+// Wait waits for the process to exit. If the process has already exited returns
+// the previous error (if any).
+func (p *JobProcess) Wait() error {
+	<-p.waitBlock
+	return p.waitError
+}
+
+// Start starts the job object process
+func (p *JobProcess) Start() error {
+	return p.cmd.Start()
+}
+
+// This should only be called once.
+func (p *JobProcess) waitBackground(ctx context.Context) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("waitBackground for JobProcess")
+
+	var err error
+	if p.cmd.Process == nil {
+		err = errors.New("process has not been started")
```

Comment:
This line seems to do nothing, since we immediately replace `err` below.

Reply:
Good catch. Don't even remember what this is leftover from